File: apps/desktop/src/lib/sync.ts
import { invoke } from "@tauri-apps/api/core";

/*
 * Typed invoke wrappers for the Phase 5 encryption + sync backend. All types are
 * camelCase; optional fields arrive as `null`. Command errors reject with the
 * shared { category, message } shape — parse them with parseLumaError.
 *
 * Secrets (passphrases, passwords, tokens) only ever flow *into* these calls
 * from transient form state. Nothing here returns or caches a secret.
 */

/**
 * Legacy device-local settings key for the global "include private keys in
 * sync" preference. Superseded by each vault's `shareSecrets` flag, which the
 * backend seeded from this value; it is still readable for one release so a
 * downgrade round-trips. Nothing reads it to decide what gets synced.
 */
export const SYNC_INCLUDE_PRIVATE_KEYS_KEY = "sync.includePrivateKeys";

/** Object-count breakdown returned by export/import/preview. */
export type ObjectCounts = {
  hosts: number;
  hostGroups: number;
  keyReferences: number;
  identities: number;
  terminalProfiles: number;
  snippets: number;
  settings: number;
  tombstones: number;
};

export type ConflictObjectType =
  | "host"
  | "host_group"
  | "key_reference"
  | "identity"
  | "terminal_profile"
  | "snippet"
  | "setting";

export type Conflict = {
  objectType: ConflictObjectType;
  objectId: string;
  label: string;
  /** Unix seconds, or null when unknown. */
  localUpdatedAt: number | null;
  /** Unix seconds, or null when unknown. */
  remoteUpdatedAt: number | null;
};

export type ConflictResolutionChoice = "keep-local" | "take-remote";

export type ConflictResolution = {
  objectType: ConflictObjectType;
  objectId: string;
  resolution: ConflictResolutionChoice;
};

/** Where Luma Cloud lives unless the user points at their own deployment. */
export const DEFAULT_LUMA_CLOUD_URL = "https://sync.luma.bwmp.dev";

export type SyncProvider =
  | "local-folder"
  | "webdav"
  | "github-gist"
  | "luma-cloud";

/**
 * When local changes are pushed without the user asking. "on-change" pushes
 * shortly after the last edit settles; "interval" batches pending changes onto
 * a fixed cadence and transfers nothing on a tick that finds none.
 */
export type AutoPushMode = "off" | "on-change" | "interval";

/**
 * This device's automatic sync schedule for one vault. Never synced itself: a
 * laptop that is awake all day and a phone on cellular want different answers,
 * so each device keeps its own (see migration 0022).
 */
export type AutoSyncSettings = {
  pushMode: AutoPushMode;
  /** Only meaningful when `pushMode` is "interval". */
  pushIntervalMinutes: number;
  /** How often the remote is polled for other devices' changes; 0 is off. */
  pullIntervalMinutes: number;
  pullOnStart: boolean;
  pullOnFocus: boolean;
};

/**
 * Cadences the backend accepts. Anything else is rejected rather than clamped,
 * so the picker and `AUTO_INTERVAL_CHOICES` in sync/mod.rs must agree.
 */
export const AUTO_INTERVAL_CHOICES = [5, 10, 15, 30, 60, 180, 360, 720, 1440] as const;

/** "10 minutes", "1 hour", "6 hours", "1 day". */
export function formatCadence(minutes: number): string {
  if (minutes < 60) return `${minutes} minutes`;
  if (minutes < 1440) {
    const hours = minutes / 60;
    return `${hours} hour${hours === 1 ? "" : "s"}`;
  }
  const days = minutes / 1440;
  return `${days} day${days === 1 ? "" : "s"}`;
}

/*
 * Both schedules are one dropdown each in the UI, but the backend models the
 * push side as a mode plus a cadence (the mode decides whether the cadence is
 * read at all). These two functions are the only place that mapping lives.
 */

/** The dropdown value representing a push schedule: "off", "on-change" or minutes. */
export function pushScheduleValue(auto: AutoSyncSettings): string {
  if (auto.pushMode === "interval") return String(auto.pushIntervalMinutes);
  return auto.pushMode;
}

/** Apply a push dropdown value, leaving the pull side untouched. */
export function withPushSchedule(
  auto: AutoSyncSettings,
  value: string,
): AutoSyncSettings {
  if (value === "off" || value === "on-change") {
    return { ...auto, pushMode: value };
  }
  return { ...auto, pushMode: "interval", pushIntervalMinutes: Number(value) };
}

export type SyncConfig = {
  vaultId: string;
  enabled: boolean;
  provider: SyncProvider | null;
  folderPath: string | null;
  url: string | null;
  username: string | null;
  gistId: string | null;
  cloudUrl: string | null;
  cloudSignedIn: boolean;
  /** Unix seconds of the last successful sync, or null. */
  lastSyncAt: number | null;
  lastRemoteVersion: string | null;
  passphraseSet: boolean;
  passphraseRemembered: boolean;
  /** This device's automatic schedule; defaults when sync is not configured. */
  auto: AutoSyncSettings;
};

export type SyncConfigureInput =
  | { provider: "local-folder"; folderPath: string }
  | { provider: "webdav"; url: string; username: string; password: string }
  | { provider: "github-gist"; token: string; gistId: string | null }
  | { provider: "luma-cloud"; cloudUrl: string };

export type SyncReport = {
  pulled: boolean;
  pushed: boolean;
  conflicts: Conflict[];
  upToDate: boolean;
  /** Private keys decrypted+imported during this sync (0 unless key sync is on). */
  privateKeysApplied: number;
  /** Private keys that could not be included because the keystore was locked. */
  privateKeysSkippedLocked: number;
};

/** Window event name carrying every automatic sync attempt (see sync/auto.rs). */
export const AUTO_SYNC_EVENT = "sync-auto";

/** What made the scheduler act. Every reason runs the same bidirectional sync. */
export type AutoSyncReason =
  | "startup"
  | "focus"
  | "change"
  | "push-interval"
  | "pull-interval";

export type AutoSyncEvent = {
  vaultId: string;
  reason: AutoSyncReason;
  phase: "started" | "completed" | "failed";
  report: SyncReport | null;
  errorCategory: string | null;
  errorMessage: string | null;
};

export type ExportResult = {
  path: string;
  objectCounts: ObjectCounts;
};

export type ImportPreview = {
  objectCounts: ObjectCounts;
  conflicts: Conflict[];
};

export type ImportApplyResult = {
  applied: ObjectCounts;
  keptLocal: ObjectCounts;
  conflicts: Conflict[];
  /** Private keys decrypted+imported during this import (0 unless included). */
  privateKeysApplied: number;
  /** Private keys that could not be imported because the keystore was locked. */
  privateKeysSkippedLocked: number;
};

/*
 * Every call below is scoped to one vault. Each vault has its own remote, its
 * own passphrase and its own baseline, so the id is a required argument rather
 * than an ambient default — a sync aimed at the wrong vault would push one
 * team's data under another's key.
 */

// Export / import -----------------------------------------------------------

export function exportEncrypted(
  vaultId: string,
  path: string,
  passphrase: string,
): Promise<ExportResult> {
  return invoke<ExportResult>("export_encrypted", { vaultId, path, passphrase });
}

export function importPreview(
  vaultId: string,
  path: string,
  passphrase: string,
): Promise<ImportPreview> {
  return invoke<ImportPreview>("import_preview", { vaultId, path, passphrase });
}

export function importApply(
  vaultId: string,
  path: string,
  passphrase: string,
  resolutions: ConflictResolution[],
): Promise<ImportApplyResult> {
  return invoke<ImportApplyResult>("import_apply", {
    vaultId,
    path,
    passphrase,
    resolutions,
  });
}

// Sync ----------------------------------------------------------------------

export function syncGetConfig(vaultId: string): Promise<SyncConfig> {
  return invoke<SyncConfig>("sync_get_config", { vaultId });
}

/** Every vault's configuration in one round trip, for the title-bar aggregate. */
export function syncListConfigs(): Promise<SyncConfig[]> {
  return invoke<SyncConfig[]>("sync_list_configs", {});
}

export function syncConfigure(vaultId: string, input: SyncConfigureInput): Promise<null> {
  return invoke<null>("sync_configure", { vaultId, input });
}

/** Replace this device's automatic schedule for one vault. */
export function syncSetAuto(
  vaultId: string,
  settings: AutoSyncSettings,
): Promise<null> {
  return invoke<null>("sync_set_auto", { vaultId, settings });
}

/**
 * Tell the backend scheduler the app is in the foreground again. Vaults with
 * `pullOnFocus` pull once, subject to the same cooldown, conflict and key
 * checks as any other automatic sync — so calling this liberally is safe.
 */
export function syncAutoFocus(): Promise<null> {
  return invoke<null>("sync_auto_focus", {});
}

export function syncSetPassphrase(
  vaultId: string,
  passphrase: string,
  remember: boolean,
): Promise<null> {
  return invoke<null>("sync_set_passphrase", { vaultId, passphrase, remember });
}

export function syncDisable(vaultId: string): Promise<null> {
  return invoke<null>("sync_disable", { vaultId });
}

export function syncNow(vaultId: string): Promise<SyncReport> {
  return invoke<SyncReport>("sync_now", { vaultId });
}

export function syncResolve(
  vaultId: string,
  resolutions: ConflictResolution[],
): Promise<SyncReport> {
  return invoke<SyncReport>("sync_resolve", { vaultId, resolutions });
}

/** Sum every object count into a single total (for compact summaries). */
export function totalObjectCount(counts: ObjectCounts): number {
  return (
    counts.hosts +
    counts.hostGroups +
    counts.keyReferences +
    counts.identities +
    counts.terminalProfiles +
    counts.snippets +
    counts.settings +
    counts.tombstones
  );
}

/**
 * Format a unix-seconds timestamp as a coarse relative string ("5 minutes
 * ago"). Defensive against null / non-finite inputs — no date library.
 */
export function formatRelativeTime(unixSeconds: number | null | undefined): string {
  if (unixSeconds == null || !Number.isFinite(unixSeconds)) return "never";
  const deltaMs = Date.now() - unixSeconds * 1000;
  if (deltaMs < 0) return "just now";
  const seconds = Math.floor(deltaMs / 1000);
  if (seconds < 45) return "just now";
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes} minute${minutes === 1 ? "" : "s"} ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours} hour${hours === 1 ? "" : "s"} ago`;
  const days = Math.floor(hours / 24);
  if (days < 30) return `${days} day${days === 1 ? "" : "s"} ago`;
  const months = Math.floor(days / 30);
  if (months < 12) return `${months} month${months === 1 ? "" : "s"} ago`;
  const years = Math.floor(months / 12);
  return `${years} year${years === 1 ? "" : "s"} ago`;
}

/** Truncate a remote-version string for compact display. */
export function truncateVersion(version: string | null | undefined, max = 12): string {
  if (!version) return "—";
  return version.length > max ? `${version.slice(0, max)}…` : version;
}

/** Human labels for sync conflict object types (singular). */
export const CONFLICT_TYPE_LABELS: Record<ConflictObjectType, string> = {
  host: "Host",
  host_group: "Host group",
  key_reference: "Key reference",
  identity: "Identity",
  terminal_profile: "Terminal profile",
  snippet: "Snippet",
  setting: "Setting",
};

File: apps/desktop/src-tauri/src/commands/sync.rs
use tauri::State;

use crate::collaboration::CollaborationRuntimeState;
use crate::errors::Result;
use crate::keystore::KeystoreState;
use crate::storage::vaults::default_id;
use crate::sync::auto::AutoSyncState;
use crate::sync::{
    self, AutoSyncSettings, ConflictResolution, ExportSummary, ImportPreview, ImportSummary,
    SyncConfig, SyncConfigureInput, SyncReport, SyncRuntimeState,
};
use crate::AppState;

#[tauri::command]
pub async fn export_encrypted(
    state: State<'_, AppState>,
    keystore_state: State<'_, KeystoreState>,
    vault_id: Option<String>,
    path: String,
    passphrase: String,
) -> Result<ExportSummary> {
    sync::export_encrypted(
        &state.pool,
        &keystore_state,
        &state.app_data_dir,
        &vault_id.unwrap_or_else(default_id),
        &path,
        &passphrase,
    )
    .await
}

#[tauri::command]
pub async fn import_preview(
    state: State<'_, AppState>,
    vault_id: Option<String>,
    path: String,
    passphrase: String,
) -> Result<ImportPreview> {
    sync::import_preview(
        &state.pool,
        &state.app_data_dir,
        &vault_id.unwrap_or_else(default_id),
        &path,
        &passphrase,
    )
    .await
}

#[tauri::command]
pub async fn import_apply(
    state: State<'_, AppState>,
    keystore_state: State<'_, KeystoreState>,
    vault_id: Option<String>,
    path: String,
    passphrase: String,
    resolutions: Vec<ConflictResolution>,
) -> Result<ImportSummary> {
    sync::import_apply(
        &state.pool,
        &keystore_state,
        &state.app_data_dir,
        &vault_id.unwrap_or_else(default_id),
        &path,
        &passphrase,
        &resolutions,
    )
    .await
}

#[tauri::command]
pub async fn sync_get_config(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
    vault_id: Option<String>,
) -> Result<SyncConfig> {
    sync::get_config(
        &state.pool,
        &runtime,
        &keystore_state,
        &vault_id.unwrap_or_else(default_id),
    )
    .await
}

#[tauri::command]
pub async fn sync_list_configs(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
) -> Result<Vec<SyncConfig>> {
    sync::list_configs(&state.pool, &runtime, &keystore_state).await
}

#[tauri::command]
pub async fn sync_configure(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
    vault_id: Option<String>,
    input: SyncConfigureInput,
) -> Result<()> {
    sync::configure(
        &state.pool,
        &runtime,
        &keystore_state,
        &state.app_data_dir,
        &vault_id.unwrap_or_else(default_id),
        input,
    )
    .await
}

/// Replace this device's automatic sync schedule for one vault. The scheduler
/// reads the row on its next tick, so there is nothing to restart.
#[tauri::command]
pub async fn sync_set_auto(
    state: State<'_, AppState>,
    vault_id: Option<String>,
    settings: AutoSyncSettings,
) -> Result<()> {
    sync::set_auto_settings(&state.pool, &vault_id.unwrap_or_else(default_id), settings).await
}

/// Tell the scheduler the app came back to the foreground. Vaults with
/// "sync when I come back" on pull once, subject to the same cooldown,
/// conflict and key checks as any other automatic sync.
#[tauri::command]
pub fn sync_auto_focus(auto: State<'_, AutoSyncState>) {
    auto.request_focus_sync();
}

#[tauri::command]
pub async fn sync_set_passphrase(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
    vault_id: Option<String>,
    passphrase: String,
    remember: bool,
) -> Result<()> {
    sync::set_passphrase(
        &state.pool,
        &runtime,
        &keystore_state,
        &vault_id.unwrap_or_else(default_id),
        passphrase,
        remember,
    )
    .await
}

#[tauri::command]
pub async fn sync_disable(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
    vault_id: Option<String>,
) -> Result<()> {
    sync::disable(
        &state.pool,
        &runtime,
        &keystore_state,
        &vault_id.unwrap_or_else(default_id),
    )
    .await
}

#[tauri::command]
pub async fn sync_now(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
    collab_runtime: State<'_, CollaborationRuntimeState>,
    vault_id: Option<String>,
) -> Result<SyncReport> {
    sync::sync_now(
        &state.pool,
        &runtime,
        &keystore_state,
        &collab_runtime,
        &state.app_data_dir,
        &vault_id.unwrap_or_else(default_id),
    )
    .await
}

#[tauri::command]
pub async fn sync_resolve(
    state: State<'_, AppState>,
    runtime: State<'_, SyncRuntimeState>,
    keystore_state: State<'_, KeystoreState>,
    collab_runtime: State<'_, CollaborationRuntimeState>,
    vault_id: Option<String>,
    resolutions: Vec<ConflictResolution>,
) -> Result<SyncReport> {
    sync::sync_resolve(
        &state.pool,
        &runtime,
        &keystore_state,
        &collab_runtime,
        &state.app_data_dir,
        &vault_id.unwrap_or_else(default_id),
        &resolutions,
    )
    .await
}
